overlord/snapstate/handlers_components.go

```diff
@@ -724,6 +724,42 @@ func (m *SnapManager) doPrepareKernelModulesComponents(t *state.Task, _ *tomb.To
 	// Inject fault during the kernel components setup
 	osutil.MaybeInjectFault("prepare-kernel-components")
 
+	var newInfo *snap.Info
+	// Set the default to false for compatibility with older snapd (case of
+	// joint refresh of snapd and kernel).
+	prepareRestart := false
+	if err := t.Get("prepare-restart", &prepareRestart); err != nil &&
+		!errors.Is(err, state.ErrNoState) {
+		return err
+	}
+	if prepareRestart {
+		// TODO we have to revert changes in bootloader config/modeenv if an
+		// error happens later in this method. This is not likely as possible
+		// errors after this would happen only due to internal errors or not
+		// being able to write to the filesystem, but still. There is also the
+		// question of what would happen if a restart happens when the boot
+		// configuration has been already written but DoneStatus in the state
+		// has not.
```

Comment:
are we happy with the reboot placement on the undo paths?

Reply:
I think the reboot placement along undo still is correct